elastic/logstash
# Implement DLQ age retention policy

## Release notes

Added age retention policy for messages inserted in DLQ. After a configurable period of time, expired messages are considered for deletion.

## What does this PR do?

Updates DLQ writer's `writeEvent` method to clean the tail segments  older then the duration period. This happens only if setting `dead_letter_queue.retain.age` is configured.
To read the age of a segment it extract the timestamp of the last (youngest) message in the segment.
The age is defined as a `number` followed by one of `d`, `h`, `m`, `s` that stands for days, hours, minutes and seconds. If nothing is used then assumes seconds as default measure entity.

## Why is it important/What is the impact to the user?

Once configured `dead_letter_queue.retain.age`, permit to the user to discard events older a certain time boundary, without processing them.

## Checklist




## Author's Checklist


- [x]  execute a real test as described in the next section

## How to test this PR locally

To test it need an Elasticsearch instance the return an error code that trigger DLQ, the DLQ feature needs to be enabled in `logstash.yml`.
- prepare the ES index to reject the inserts, create an index and close it:
```
PUT test_index/
POST test_index/_close
```
-  in `logstash.yml` enable this DLQ feature, say the events should lasts for 2 minutes:
```
dead_letter_queue.enable: true
dead_letter_queue.retain.age: 2m
```
- run a pipeline that produces events into DLQ:
```
input {
  generator {
    # this is some data
    message => '{"jvm" : {"threads" : {"count" : 49,"peak_count" : 50},"mem" : {"heap_used_percent" : 14,"heap_committed_in_bytes" : 309866496,"heap_max_in_bytes" : 1037959168,"heap_used_in_bytes" : 151686096,"non_heap_used_in_bytes" : 122486176,"non_heap_committed_in_bytes" : 133222400,"pools" : {"survivor" : {"peak_used_in_bytes" : 8912896,"used_in_bytes" : 288776,"peak_max_in_bytes" : 35782656,"max_in_bytes" : 35782656,"committed_in_bytes" : 8912896},"peak_used_in_bytes" : 148656848,"used_in_bytes" : 148656848,"peak_max_in_bytes" : 715849728,"max_in_bytes" : 715849728,"committed_in_bytes" : 229322752},"young" : {"peak_used_in_bytes" : 71630848,"used_in_bytes" : 2740472,"peak_max_in_bytes" : 286326784,"max_in_bytes" : 286326784,"committed_in_bytes" : 71630848}}},"gc" : {"collectors" : {"old" : {"collection_time_in_millis" : 607,"collection_count" : 12},"young" : {"collection_time_in_millis" : 4904,"collection_count" : 1033}}},"uptime_in_millis" : 1809643}'
    codec => json
  }
}

output {
  elasticsearch {
    index => "test_index"
    hosts => "http://localhost:9200"
    user => "elastic"
    password => "changeme"
  }
}
```
- after launched Logstash, verify that the tail dlq segments starts to be removed  after the 2 minutes period:
```sh
ll data/dead_letter_queue/dlq_upstream
```

## Related issues


- Superseeds #13808
- Relates #10493

## Use cases

As user with DLQ  enabled I want that events older a certain duration are expired and not processed, the space occupied must be freed.
